File: 2021/go/day1.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func Part1(input []int) int {

	max := input[0]
	count := 0
	for i := 1; i < len(input); i++ {
		num := input[i]
		if num > max {
			count++
		}
		max = num
	}
	return count
}

func Part2(input []int) int {
	startQueue := []int{}
	result := []int{}
	for i := 0; i < 3; i++ {
		startQueue = append(startQueue, input[i])
	}
	val := sum(startQueue)
	result = append(result, val)

	for i := 3; i < len(input); i++ {
		val = val - input[i-3] + input[i]
		// startQueue = append(startQueue[1:], input[i])
		result = append(result, val)
	}

	return Part1(result)

}

func sum(input []int) int {
	sum := 0
	for _, num := range input {
		sum += num
	}
	return sum
}

func main() {
	text, err := os.ReadFile("./../input/day1.txt")
	if err != nil {
		fmt.Println(err)
		return
	}
	input := strings.Split(string(text), "\n")
	inputInt := []int{}

	for _, line := range input {
		num, _ := strconv.Atoi(line)
		inputInt = append(inputInt, num)
	}

	fmt.Println(Part1(inputInt))
	fmt.Println(Part2(inputInt))
}

File: 2021/js/day1.js
let fs = require("fs");
let path = require("path");

let input = fs
  .readFileSync(path.resolve(__dirname, "./../input/day1.txt"), "utf8")
  .split("\n");

const part1 = (arr) => {
  let max = Number(arr[0]);
  let count = 0;
  for (let i = 1; i < arr.length; i++) {
    let num = Number(arr[i]);
    if (num > max) {
      count++;
    }
    max = num;
  }
  return count;
};

const part2 = (arr) => {
  const queue = arr.slice(0, 3);
  const result = [queue.reduce((a, b) => Number(a) + Number(b))];
  let val = queue.reduce((a, b) => Number(a) + Number(b));
  for (let i = 3; i < arr.length; i++) {
    val = val - Number(arr[i - 3]) + Number(arr[i]);
    // result.push(maintainQueue(queue, arr[i]));
    result.push(val);
  }
  return part1(result);

  function maintainQueue(arr, el) {
    arr.push(el);
    arr.shift();
    return arr.reduce((a, b) => Number(a) + Number(b));
  }
};

console.log(part1(input));
console.log(part2(input));
